visualization/app/codeCharta/ui/attributeSideBar/attributeSideBarSecondaryMetrics/secondaryMetrics.selector.ts

```diff
@@ -0,0 +1,28 @@
+import { createSelector } from "../../../state/angular-redux/store"
+import { selectedNodeSelector } from "../../../state/selectors/selectedNode.selector"
+import { CcState } from "../../../state/store/store"
+import { primaryMetricNamesSelector } from "../attributeSideBarPrimaryMetrics/primaryMetricNames.selector"
+import { Metric, shouldShowAttributeTypeSelector } from "../util/metricHelper"
+
+export type SecondaryMetric = Metric & {
+	showDeltaValue: boolean
+}
+
+export const secondaryMetricsSelector: (state: CcState) => SecondaryMetric[] = createSelector(
+	[selectedNodeSelector, primaryMetricNamesSelector],
+	(selectedNode, primaryMetricNames) => {
+		if (!selectedNode) return []
```

Comment:
secondary metrics are calculated by node of selected building. If there is no building selected in the map, there is no `selectedNode`

Reply:
I understood that. I would just expect the secondary metrics to only be required if a building is selected?

Reply:
yes it is. But I strongly think a selector should never throw if given a valid store state. As selected building can be undefined in store, I think we should handle it here as well, regardless if the selector will only be called within the sidebar, which should only be rendered if there is a node.

Reply:
I understand your point. Yet, what would the receiving side do in case it receives an empty array? It might be able to handle it gracefully but maybe it does not. As such, let's go for a middle ground: throw an error in case it's called without selected node. That way we are able to let our selves be reminded how to use the functionality correct and the receiving side is never going to receive things that it's not able to handle. WDYT?

Reply:
I would be OKish with adding jsdocs for throwing, or throwing an explicit exception. Although I still think selectors should be able to handle every possible state gracefully.

Besides an empty list is a valid secondary metric. A map could consist of only 3 metrics which are all primary metrics. Therefore every component using secondary metrics have to deal with that case.

Anyway I just tried out what happens when not doing this check (this nullish check of node was also done in none migrated implementation). Currently selecting another building results in deselecting first (which crashes the side bar without the check) and then selecting the new building. So lets be save with the check for now.